### PR TITLE
Refactoring well test to prepare for Physical limit related well test. 

### PR DESCRIPTION
**Pull request by GitPaean**
Re-factoring the well test related functionalities to prepare for the physical limits well test functionalities. 

Well can be shut/closed and re-opened due to reason other than economic reasons. The focus at the moment is PHYSICAL limit.

Nothing in simulation result should be changed by this PR.

### CODE REVIEW

**Comment by GitPaean** [comment]
jenkins build this please

File: opm/autodiff/WellInterface_impl.hpp [hunk 1]
```diff
@@ -641,30 +641,72 @@ namespace Opm
...
+
+    template<typename TypeTag>
+    void
+    WellInterface<TypeTag>::
+    updateWellTestStatePhysical(const WellState& /* well_state */,
+                                const double /* simulation_time */,
+                                const bool /* write_message_to_opmlog */,
+                                WellTestState& /* well_test_state */) const
+    {
+        // TODO: other facilities are required to make the following action work
```
Comment (GitPaean) [inline opm/autodiff/WellInterface_impl.hpp]: This PR is created based on the WIP branch. I did not realize other facilities are required for the `updateWellTestStatePhysical`, which makes this function an empty function.  Please let me know if you think I should remove this function from this PR.

**Comment by GitPaean** [comment]
The purpose of this PR is to separate the closing/re-opening well due to different reasons.  The interaction between different reasons can be tricky and I do not have a answer now.  It has to be addressed when we have a case that involving that.

**Comment by GitPaean** [comment]
jenkins build this please

**Comment by totto82** [comment]
It is hard to judge this PR based on it alone without the "updateWellTestStatePhysical" part. Currently this PR just complicates the well testing without any good reason. Why is this refactoring needed? In my naive mind I thought updating the well testing with a new condition would more or less be trivial? Maybe you could push the branch where the PHYSICAL reason is included.

**Comment by GitPaean** [comment]
> It is hard to judge this PR based on it alone without the "updateWellTestStatePhysical" part. Currently this PR just complicates the well testing without any good reason. Why is this refactoring needed? In my naive mind I thought updating the well testing with a new condition would more or less be trivial? Maybe you could push the branch where the PHYSICAL reason is included.

Because, the way of testing PHYSICAL reason is very different from ECONOMICAL reason. Testing with PHYSICAL reason might not result in any result, while the way to test ECONOMICAL reason need to run the simulation and get converged result then check the `WellState`.  This PR is trying to separate the well testing for different reasons for now to avoid ruining the currently accomplished functionality (ECONOMIC limit well testing) when developing well testing for PHYSICAL reasons. 

The current PR tries to reduce complexity and length of the function `wellTesting` in `BlackoilWellModel`, and moving the well testing to happen within each individual well. And the functions are shorter and things are more local to each individual well. 

Unless a well can be shut because of two different reasons (it means the well is shut at the same time for PHYSICAL and ECONOMICAL reasons) and need to be tested for two different reasons, the approach suggested by this PR is a better approach. 

And, even the above condition happens, it is still good to move the `well testing` to happen in each individual well (`WellInterface`) instead of doing it in the `BlackoilWellModel`.

**Comment by GitPaean** [comment]
> I think as a minimum we should add a simple test where both phycial and economical limites are used. In fact Kjersti in Equinor gave me such a case earlier.

It is for sure good to do. The way of handling ECONOMICAL limits is relatively clear, while they way of handling  PHYSICAL limits is still under development.  I believe it is good to separate these two for now and test and develop them separately.  When we have both work properly, we put them together. Since we understand and handle both in a reasonable way, the efforts that combine both together will be for good purposes and easier to handle. 

And also, we are trying to only introduce only small PRs for better PR review (unless no way to split the PR).

**Comment by totto82** [comment]
Thanks for the patch. Moving the well testing from the BlackoilWellModel to the well interface class seems like a good thing. Since all tests are green, I will just merge it as is.

**Pull request merged**